cargo/private/cargo_build_script.bzl

```diff
@@ -225,7 +225,7 @@ def _cargo_build_script_impl(ctx):
         streams.stderr.path,
     ])
     build_script_inputs = []
-    for dep in ctx.attr.deps:
+    for dep in ctx.attr.link_deps:
```

Comment:
I believe we still want to include the `out_dir` stuff for _all_ deps, but only gate the `dep_env` adding based on `links`? (I think this will be simpler with a new provider field - we can just iterate across all `deps`, but conditionally do the `dep_env` stuff based on the presence of the `links` field in the provider we're inspecting)

Reply:
I think I may have screwed this part up. I'll take a look at it again some time soon.

Reply:
Okay, after double checking how cargo works, it doesn't pay attention to `cargo:rerun-if-changed` for things in the `OUT_DIR` so the only option there is to rely on `cargo:rerun-if-env-changed` which can only come from `link_deps` assuming that the cargo build.rs wants to be correct (admittedly some build.rs scripts do stupid things (such as `pyo3`, but manually applying fixes to make those work at all in Bazel is almost always required so I think that the few extra lines of code there is worth the globally reduced build times elsewhere). Therefore I think in theory we only need to make available the parts of `OUT_DIR` from a `cargo_build_script`s `link_deps` and the `OUT_DIR`s that those `link_deps` can see (as you can keep passing forward paths by repeated uses of `cargo:KEY=VALUE`). In practice this also seems to be what I am observing.

If you agree with this conclusion I will rework line 233:
`            for dep_build_info in dep[rust_common.dep_info].transitive_build_infos.to_list():`
to only be for transitive `link_deps`.

Reply:
I have decided not to make this change, but just to leave a comment about it. The code was hard for me to understand that handled the current `transitive_build_infos` due to the codes indirectness, lack of comments, and lack of types. I spent an hour or so staring at the screen feeling a bit foolish before giving up. I may try again later, but I believe that making the change will just minorly improve build caching.

Reply:
I thought about this more, and made some examples in `cargo` showing how this whole set up is way more complicated than I initially realized. I have switched it to be more like what you asked. OUT_DIR can basically come from anywhere I realized. Build dependencies, link dependencies, regular dependencies, proc macro dependencies, etc. They only "work" in that they assume that the cargo target directory doesn't change at all during the build process which is true unless you `cargo build`, and then move the whole repository, and then `cargo build` again.

I think that this isn't actually possible to handle properly with poorly behaving cargo crates. I think I may try to contribute a cargo clippy lint for this bad behavior in the next weeks, and maybe rules_rust should consider supporting those poorly behaving crates out of scope. I'm not sure what the right decision is.

[foobar.tar.gz](https://github.com/bazelbuild/rules_rust/files/11826078/foobar.tar.gz)